tests/robot-tests/README.md

```diff
@@ -228,8 +228,186 @@ There is a known bug with pyderman which is responsible for downloading chromedr
+pipenv run python run_tests.py -f tests/visual_testing/visually_check_tables_and_charts.dev.robot -e dev --rerun-failed-tests
+```
+
+#### Visually compare images
+Now that we have a set of "before" and "after" snapshot images in the two `snapshots-<yyyy-MM-dd hh:mm:ss>` folders, we're able to visually compare them to 
+ensure that only expected changes are present.
+
+From the `<project root>/useful-scripts/compare_image_snapshots` folder, run:
+
+```bash
+pipenv run python compare_image_folder_trees.py --first </path/to/before/snapshots> --second </path/to/after/snapshots> --diff </path/to/diffs/folder>
+```
+
+This will compare the images in the `</path/to/before/snapshots>` folder with the images in the `</path/to/after/snapshots>` folder.
+
+Any differences will be output to the provided `</path/to/diffs/snapshots>` folder. This folder will be created if it does not already exist. If 
+differences are found, the original, the new, and the diff images are all output to allow manual visual comparisons.
+
+## Permalinks
+The process for visually testing Permalinks is the same as with tables and charts above, with the exceptions that the CSV that drives the test is different, 
+and that the command to generate the test script is slightly different.
+
+### CSV
+The CSV for Permalink testing is a single column of Permalink Ids. For each Id, a test case will be generated. There's no process for generating a CSV of Permalink 
+Ids for an environment at this point in time whilst they remain in Table Storage, but known Ids can be supplied in a `permalinks-dev.csv` as an example for the Dev
+environment.
+
+### Test script generation
+To generate a test script for an environment, e.g. Dev environment using the CSV above, run:
+
+```bash
+cd tests/robot-tests/tests/visual_testing
+pipenv run python generate_permalink_test_cases.py --file /path/to/tests/robot-tests/permalinks-dev.csv --target visually_check_permalinks.dev.robot 
+```
+
+This test script can then be run for befores and afters and then images visually compared as per the tables and charts example. The command for running these tests 
+for the Dev environment example would be:
+
+```bash
+pipenv run python run_tests.py -f tests/visual_testing/visually_check_permalinks.dev.robot -e dev
+```
+
+## Limitations of visual testing
+Currently there are a few limitations with this process:
+
+1. Comparing images of different dimensions - the `compare_image_folder_trees.py` script will attempt to resize images prior to comparing. Currently there are 
+   occasional errors whilst resizing images of particular dimensions.
+2. Occasional capture of a table and chart before all styling has been applied - this will lead to visual differences due to, for instance, the correct font not 
+   being applied yet, due to a flash of unstyled content.
+3. Tables being too large for image capture - in some environments (prior to the maximum table cell limit being applied), some tables are so huge that Selenium
```

Comment:
Not too sure if you reckon we can do a scrolling screenshot (i.e. take a screenshot of the table currently in the viewport, scroll until the previous table cells are out of the viewport & repeat the process until the entire table is capture + stitch them together as a 1 whole image of the table). I reckon this might be quite complex so this'd be only worth it if it provides enough value

Reply:
Yup agreed yeah it'd be nice to do - I think there might be other libs we might be able to call on to help with this too - I just can't find their name right now but I came across one quite frequently.